README.md
# gcxh-js-ol-esay

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.3.4](https://gitee.com/SatImage2014/gcxh-js-ol-esay/compare/v0.3.3...v0.3.4) (2022-08-25)


### Features

* ol版本升级 ([18cf957](https://gitee.com/SatImage2014/gcxh-js-ol-esay/commit/18cf957142ab3f3566a8a9a4cc48d0e9a9144c4e))

### [0.3.3](https://gitee.com/SatImage2014/gcxh-js-ol-esay/compare/v0.3.2...v0.3.3) (2022-07-04)


### Features

* 支持添加 `ArcGIS Server Layer` ([9f78d99](https://gitee.com/SatImage2014/gcxh-js-ol-esay/commit/9f78d9929065eee8db63e46a872d9cf18f2922b6))

### [0.3.2](https://gitee.com/SatImage2014/gcxh-js-ol-esay/compare/v0.3.1...v0.3.2) (2022-06-07)


### Bug Fixes

* 添加天地图-全球影像地图服务（离线部署） ([a5778de](https://gitee.com/SatImage2014/gcxh-js-ol-esay/commit/a5778dead2de5a5fd12335c728ab1096a8ac7cfc))

### [0.3.1](https://gitee.com/SatImage2014/gcxh-js-ol-esay/compare/v0.3.0...v0.3.1) (2022-03-16)


### Bug Fixes

* 传入 `extent` 设定为 `EPSG:4326` 坐标系 ([e6d3a00](https://gitee.com/SatImage2014/gcxh-js-ol-esay/commit/e6d3a004a54de01e351a7ecd9db5ad52f31aadcf))

## 0.3.0 (2022-03-15)


### Features

* **init:** init(暂时关闭eslint) ([bdc1a14](https://gitee.com/SatImage2014/gcxh-js-ol-esay/commit/bdc1a14d33b451b72c4ddf1603247d796fec7964))
* 全面升级 ([25edd0e](https://gitee.com/SatImage2014/gcxh-js-ol-esay/commit/25edd0ebd4a7b43295100263eafda0f9bcf28f6c))


lib/Layer.js
/*
 * @Author       : Chen Zhen
 * @Date         : 2018-10-27 15:35:32
 * @LastEditors  : Han Jia Yi
 * @LastEditTime : 2022-06-07 14:27:16
 * @Description  : 基础地图底图图层类
 */

import _ from 'lodash'

import TiandituLayer from './TiandituLayer'

import GCXHLayer from './GCXHLayer'

const KNOW_LAYERS = [
  {
    name: 'TiandituLayerVecW',
    label: '天地图-全球矢量地图服务',
    layer: ({ tiandituToken }) => (new TiandituLayer({
      id: 'TiandituLayerVecW',
      type: '矢量底图',
      token: tiandituToken,
    })),
  },
  {
    name: 'TiandituLayerVcaW',
    label: '天地图-全球矢量中文注记服务',
    layer: ({ tiandituToken }) => (new TiandituLayer({
      id: 'TiandituLayerVcaW',
      type: '矢量标注',
      token: tiandituToken,
    })),
  },
  {
    name: 'TiandituLayerImgW',
    label: '天地图-全球影像地图服务',
    layer: ({ tiandituToken }) => (new TiandituLayer({
      id: 'TiandituLayerImgW',
      type: '影像底图',
      token: tiandituToken,
    })),
  },
  {
    name: 'TiandituLayerCiaW',
    label: '天地图-全球影像中文注记服务',
    layer: ({ tiandituToken }) => (new TiandituLayer({
      id: 'TiandituLayerCiaW',
      type: '影像标注',
      token: tiandituToken,
    })),
  },
  {
    name: 'TiandituLayerTerW',
    label: '天地图-全球地形晕渲地图服务',
    layer: ({ tiandituToken }) => (new TiandituLayer({
      id: 'TiandituLayerTerW',
      type: '地形底图',
      token: tiandituToken,
    })),
  },
  {
    name: 'TiandituLayerCtaW',
    label: '天地图-全球地形中文标记服务',
    layer: ({ tiandituToken }) => (new TiandituLayer({
      id: 'TiandituLayerCtaW',
      type: '地形标注',
      token: tiandituToken,
    })),
  },
  {
    name: 'TiandituLayerIboW',
    label: '天地图-全球行政边界服务（影像）',
    layer: ({ tiandituToken }) => (new TiandituLayer({
      id: 'TiandituLayerIboW',
      type: '边界-影像',
      token: tiandituToken,
    })),
  },
  {
    name: 'TiandituLayerTboW',
    label: '天地图-全球行政边界服务（地形）',
    layer: ({ tiandituToken }) => (new TiandituLayer({
      id: 'TiandituLayerTboW',
      type: '边界-地形',
      token: tiandituToken,
    })),
  },

  {
    name: 'InnerTiandituLayerVecC',
    label: '天地图-全球矢量地图服务（离线部署）',
    layer: ({ url, layer }) => (new GCXHLayer({
      id: 'InnerTiandituLayerVecC',

      url,
      layer: layer || 'vec_c',
    })),
  },
  {
    name: 'InnerTiandituLayerCiaC',
    label: '天地图-全球矢量中文注记服务（离线部署）',
    layer: ({ url, layer }) => (new GCXHLayer({
      id: 'InnerTiandituLayerCiaC',

      url,
      layer: layer || 'cia_c',
    })),
  },
  {
    name: 'InnerTiandituLayerTile',
    label: '天地图-全球影像地图服务（离线部署）',
    layer: ({ url, layer }) => (new GCXHLayer({
      id: 'InnerTiandituLayerTile',

      url,
      layer: layer || 'tile_c',
    })),
  },
]

// let mapListNameCache = null // 用于懒缓存地图的name属性

/**
 * 支持的地图底图的图层数据，可以进行全局获取，亦可以根据对应的编码进行获取
 */
class Layer {
  /**
   * 返回支持的地图图层的数据
   */
  static KNOW_LAYERS() {
    return _.cloneDeep(KNOW_LAYERS)
  }

  // /**
  //  * 返回支持的底图对应的名称（name属性）
  //  */
  // static MapListName() {
  //   if (_.isNull(mapListNameCache)) {
  //     mapListNameCache = MAP_LIST.map((i) => i.name)
  //   }

  //   return mapListNameCache
  // }

  /**
   * 获取某个基础地图底图
   *
   * @param {String|Object} baselayer 基础地图底图的名称
   *
   * @param {Object} defaultOption MapPlugins提供的默认配置
   *
   * @return {Layer | null} 返回的底图
   *
   */
  static parse(baselayer, defaultOption) {
    if (_.isArray(baselayer)) {
      const layers = baselayer.map((i) => Layer.parse(i, defaultOption))

      return layers.filter((i) => i)
    }

    if (_.isString(baselayer)) {
      const baseLayer = KNOW_LAYERS.find((i) => i.name === baselayer)

      if (baseLayer) return baseLayer.layer(defaultOption)
      return null
    }

    if (_.isObject(baselayer)) {
      if ([ 'Tianditu', 'InnerTiandituGCXH' ].includes(baselayer.type)) {
        const baseLayer = KNOW_LAYERS.find((i) => i.name === baselayer.layer.name)
        if (baseLayer) return baseLayer.layer({ ...defaultOption, ...baselayer.layer })
        return null
      }

      if (baselayer.type === 'WMTS') {
        // ...
        return null
      }

      if (baselayer.type === 'custom') {
        return baselayer.layer
      }

      throw new TypeError(`option: ${baselayer} id not object or not string`)
    }

    throw new TypeError(`option: ${baselayer} id not object or not string`)
  }
}

export default Layer


lib/GCXHLayer.js
/*
 * @Author       : Chen Zhen
 * @Date         : 2020-12-16 19:54:10
 * @LastEditors  : Chen Zhen
 * @LastEditTime : 2020-12-16 20:01:32
 */

import TileLayer from 'ol/layer/Tile'

import WMTS from 'ol/source/WMTS'
import WMTSTileGrid from 'ol/tilegrid/WMTS'
import Projection from 'ol/proj/Projection'

import {
  get as getProjection,
} from 'ol/proj'

import { getWidth } from 'ol/extent'

const DEFAULT_OPTION = {
  type: '影像底图',
  opacity: 1,
  source: undefined,
  visible: true,
  extent: undefined,
}

const map = new Map()

const getTileInfo = (projection) => {
  if (!map.has(projection)) {
    const matrixIds = []
    const resolutions = []

    const tileSizePixels = 256
    const tileSizeMtrs = getWidth(getProjection(projection).getExtent()) / tileSizePixels

    for (let i = 0; i <= 16; i += 1) {
      matrixIds[i] = i
      resolutions[i] = tileSizeMtrs / (2 ** (i + 1))
    }

    map.set(projection, {
      matrixIds,
      resolutions,
    })
  }
  return map.get(projection)
}

class GCXHLayer {
  constructor(option) {
    const {
      id,
      opacity,
      url,
      layer,
      // type,
      visible = true,
      extent = [ 0, 0.0, 180.0, 90.0 ],
      projection = 'EPSG:4326',
    } = Object.assign(DEFAULT_OPTION, option)

    const {
      matrixIds,
      resolutions,
    } = getTileInfo(projection)

    const source = new WMTS({
      // 服务地址
      url,
      // url: `http://10.36.93.240:7090/rest/wmts`,
      layer,
      // 切片集
      matrixSet: 'satImage',
      format: 'image/jpeg',
      projection: new Projection({
        code: 'EPSG:4326', // 投影编码
        units: 'degrees',
        axisOrientation: 'neu',
      }),

      tileGrid: new WMTSTileGrid({
        tileSize: [ 256, 256 ],
        extent, // 范围
        origin: [ -180.0, 90.0 ],
        resolutions,
        matrixIds,
      }),
      wrapX: true,
    })

    const tileLayer = new TileLayer({
      id,
      opacity, // 可见度
      source,
      visible,
    })

    return tileLayer
  }
}

export default GCXHLayer


lib/TiandituLayer.js
/*
 * @Author       : Chen Zhen
 * @Date         : 2020-09-22 20:18:53
 * @LastEditors  : Chen Zhen
 * @LastEditTime : 2021-12-14 14:43:59
 * @Description  : openlayer 天地图类
 *
 * 根据传入的参数,返回不同的天地图服务
 * 全部采用墨卡托的数据
 *
 */

import { getWidth, getTopLeft } from 'ol/extent'
import { get as getProjection } from 'ol/proj'

import WMTSTileGrid from 'ol/tilegrid/WMTS'
import WMTS from 'ol/source/WMTS'
import TileLayer from 'ol/layer/Tile'

let tileGrid = null
const getTileGrid = () => {
  const projection = getProjection('EPSG:3857')
  const tileSizePixels = 256
  const tileSizeMtrs = getWidth(projection.getExtent()) / tileSizePixels
  const matrixIds = []
  const resolutions = []

  for (let i = 0; i <= 18; i += 1) {
    matrixIds[i] = i
    resolutions[i] = tileSizeMtrs / (2 ** i)
  }

  return new WMTSTileGrid({
    origin: getTopLeft(projection.getExtent()),
    resolutions,
    matrixIds,
  })
}

const DEFAULT_OPTION = {
  type: '影像底图',
  opacity: 1,
  source: undefined,
  visible: true,
  extent: undefined,
}

/**
 * 根据传入的不同类型名称，
 * 返回不同的地图名称
 * @param {String} layerName 图层的名称
 */

const getLayerInfo = (layerName) => {
  switch (layerName) {
    case '影像底图':
      return 'img_w'
    case '影像标注':
      return 'cia_w'
    case '地形底图':
      return 'ter_w'
    case '地形标注':
      return 'cta_w'
    case '矢量底图':
      return 'vec_w'
    case '矢量标注':
      return 'cva_w'
    case '边界-影像':
      return 'ibo_w'
    case '边界-地形':
      return 'tbo_w'

    default:
      return 'img_w' // 默认为影像底图
  }
}

/**
 * 获取对应的urls
 * @param {String} type 地图英文名称
 */

class TiandituLayer {
  constructor(option) {
    const {
      id,
      opacity,
      type,
      visible,
      extent,
      token = null,
    } = Object.assign(DEFAULT_OPTION, option)

    if (!tileGrid) {
      tileGrid = getTileGrid()
    }

    const layerName = getLayerInfo(type)

    const layer = new TileLayer({
      id,
      name: option.id,
      opacity, // 可见度
      source: new WMTS({
        url: `http://t{0-7}.tianditu.gov.cn/${layerName}/wmts?tk=${token}`,
        layer: layerName.match(/^[a-z]*/)[0],
        format: 'tiles',
        matrixSet: 'w', // w 对应的为Web墨卡托投影
        tileGrid,
        style: 'default',
        wrapX: true,
      }),
      visible, // 是否可见
      extent, // 视野范围
    })
    return layer
  }
}

export default TiandituLayer


lib/MapPlugins.js
/*
 * @Author       : Chen Zhen
 * @Date         : 2018-10-27 14:01:47
 * @LastEditors  : Chen Zhen
 * @LastEditTime : 2022-03-16 10:00:12
 * @Description  : 地图主类
 */

import _ from 'lodash'

import 'ol/ol.css'
import Map from 'ol/Map'
import View from 'ol/View'
// import Layer from 'ol/layer/Layer'

import {
  transform,
  transformExtent,
  // get as getProjection,
} from 'ol/proj'

// import { getWidth, getTopLeft } from 'ol/extent'

import Point from 'ol/geom/Point'

import VectorLayer from 'ol/layer/Vector'
import VectorSource from 'ol/source/Vector'

import { defaults as defaultInteractions } from 'ol/interaction'

// import Projection from 'ol/proj/Projection'
// import TileWMS from 'ol/source/TileWMS'
import ImageWMSSource from 'ol/source/ImageWMS'
import ImageLayer from 'ol/layer/Image'

import TileLayer from 'ol/layer/Tile'
// import WMTS from 'ol/source/WMTS'
// import WMTSTileGrid from 'ol/tilegrid/WMTS'
import XYZ from 'ol/source/XYZ'

import { defaults as defaultControls } from 'ol/control'

import LayerPlugins from './Layer'
import DrawTool from './DrawTool'
import Measure from './Measure'

export const DEFAULT_CENTER = [ 104.29482679000004, 38.11539371468646 ]
export const DEFAULT_ZOOM = 4

// if (/(iPhone|iPad|iPod|iOS)/i.test(navigator.userAgent)) {
//   DEFAULT_ZOOM = 3.8
// } else if (/(Android)/i.test(navigator.userAgent)) {
//   DEFAULT_ZOOM = 3.8
// }

// // eslint-disable-next-line max-len
// const gridNames = [ 'EPSG:4326:0', 'EPSG:4326:1', 'EPSG:4326:2', 'EPSG:4326:3', 'EPSG:4326:4', 'EPSG:4326:5', 'EPSG:4326:6', 'EPSG:4326:7', 'EPSG:4326:8', 'EPSG:4326:9', 'EPSG:4326:10', 'EPSG:4326:11', 'EPSG:4326:12', 'EPSG:4326:13', 'EPSG:4326:14', 'EPSG:4326:15', 'EPSG:4326:16', 'EPSG:4326:17', 'EPSG:4326:18', 'EPSG:4326:19', 'EPSG:4326:20', 'EPSG:4326:21' ]

// // 切片大小
// // eslint-disable-next-line max-len
// const resolutions = [ 0.703125, 0.3515625, 0.17578125, 0.087890625, 0.0439453125, 0.02197265625, 0.010986328125, 0.0054931640625, 0.00274658203125, 0.001373291015625, 6.866455078125E-4, 3.4332275390625E-4, 1.71661376953125E-4, 8.58306884765625E-5, 4.291534423828125E-5, 2.1457672119140625E-5, 1.0728836059570312E-5, 5.364418029785156E-6, 2.682209014892578E-6, 1.341104507446289E-6, 6.705522537231445E-7, 3.3527612686157227E-7 ]

// const DEFAULT_EPSG = 'EPSG:3857'

/**
 * 解析地图容器
 * @param {HTMLElement | String} container 地图的容器
 */
const getContainer = (container) => {
  if (_.isElement(container)) return container
  if (_.isString(container)) {
    const dom = document.getElementById(container)
    if (_.isElement(dom)) return dom
    throw new ReferenceError(`container: ${container} is String, but not HTMLElement Id`)
  } else {
    throw new TypeError(`container: ${container} is not HTMLElement or String(HTMLElement Id)`)
  }
}

/**
 *
 * TODOLIST
 * 1. 增加 点 线 面 选中功能
 * 2. 增加 灵活替换底图的机制
 * 3. 增加 添加各项地图的方式 着重 wms wfs 服务 并将服务发布的方式快捷发布
 */

class MapPlus {
  /**
   *
   * 初始化地图容器的构造函数
   *
   * @param {HTMLElement | String} container 地图的容器
   * @param {Object} option 配置项
   * @param {String | Array<String | Layer>} baseLayers 底图的名称，可以支持多个图层的加载,加载顺序为，排列在前的优先进行加载，排在最下侧
   */
  constructor(container, {
    interactions = null,
    view,
    baseLayers = null, // 默认底图
    defaultExtent,
    defaultCenter = DEFAULT_CENTER,
    defaultZoom = DEFAULT_ZOOM,
    maxZoom = 18,
    defaultPluginControls = [],
    tiandituToken,

    // epsg = DEFAULT_EPSG
  } = {}) {
    this.version = '0.0.1' // 版本号

    this.tiandituToken = tiandituToken

    this.container = getContainer(container) // 地图的容器

    this.baseLayers = this.getBaselayers(baseLayers) // 地图需要进行加载的底图

    this._map = null // 对应的Map

    this._measure = null // 用来保存测量功能对象的属性

    this._drawTool = null // 用来保存绘制功能对象的属性

    this._view = view // 之前继续你功能赋值的view

    this.defaultExtent = defaultExtent ? transformExtent(defaultExtent, 'EPSG:4326', 'EPSG:3857') : null

    this.defaultCenter = transform(defaultCenter, 'EPSG:4326', 'EPSG:3857')

    this.defaultZoom = defaultZoom

    this.interactions = interactions

    this._maxZoom = maxZoom

    this.defaultPluginControls = defaultPluginControls

    this.init() // 初始化

    return this
  }

  /**
   * ------------------------------------------------
   * -------------------内部方法----------------------
   * ------------------------------------------------
   */

  /**
   * ------------------------------------------------
   * --------------------基础------------------------
   * ------------------------------------------------
   */

  getMap() {
    return this._map
  }

  /**
   * 初始化方法，不允许重复调用
   */
  init() {
    // 默认使用 EPSG:3857 坐标系
    // 如果在外部穿参 进来 则直接替换view
    const view = this._view || new View({
      // projection: new Projection(this.epsg),
      // extent: this.defaultExtent,
      center: this.defaultCenter, // 中心点设定为天安门
      zoom: this.defaultZoom,
      maxZoom: this._maxZoom || 18,
      minZoom: 1,
    })
    this._map = new Map({
      view,
      controls: defaultControls({ zoom: false }).extend(this.defaultPluginControls),

      interactions: this.interactions || defaultInteractions({
        pinchRotate: false,
        doubleClickZoom: false,
      }),

      layers: this.baseLayers, // 默认值为 null
      target: this.container,
    })

    // defaultExtent 优先级更高一点
    if (_.isArray(this.defaultExtent) && this.defaultExtent.length === 4) view.fit(this.defaultExtent, { duration: 0 })
  }

  getCenter() {
    return transform(this.getMap().getView().getCenter(), 'EPSG:3857', 'EPSG:4326')
  }

  getExtent() {
    return transformExtent(this.getMap().getView().calculateExtent(), 'EPSG:3857', 'EPSG:4326')
  }

  getZoom() {
    return this.getMap().getView().getZoom()
  }

  /**
   * 销毁方法
   */
  destory() {
    // 需要进行销毁测量功能
    if (this._measure instanceof Measure) this._measure.destory()
  }

  /**
   * resize事件需要进行调用的事件
   */
  // eslint-disable-next-line class-methods-use-this
  resizeHandle() {
    this._map.setSize([ this.container.offsetWidth, this.container.offsetHeight ])
    this._map.setSize([ this.container.offsetWidth, this.container.offsetHeight ])
  }

  /**
   * 执行ol,map的render钩子函数
   */
  // eslint-disable-next-line class-methods-use-this
  renderHandle() {}

  /**
   * 恢复初始化默认位置
   */
  moveHome() {
    const _view = this._map.getView()

    if (_.isArray(this.defaultExtent) && this.defaultExtent.length === 4) {
      _view.fit(this.defaultExtent, { duration: 0 })
    } else {
      // this._map.view.
      _view.setCenter(this.defaultCenter)
      _view.setZoom(this.defaultZoom)
    }
  }

  moveHomeFit({
    duration = 1000,
    maxZoom = this.defaultZoom,
    callback = null,
  } = {}) {
    this._map.getView().fit(new Point(this.defaultCenter), {
      duration,
      maxZoom,
      callback: () => {
        // this.spotLayer.setVisible(true)
        if (callback) callback()
      },
    })
  }

  /**
   * 增加图层
   */
  addLayer(layer, opt = {}) {
    if (_.isNil(opt.index)) {
      this._map.addLayer(layer)
    } else {
      layer.setZIndex(opt.index)
      this._map.getLayers().insertAt(opt.index, layer)
    }
  }

  /**
   * 移除对应的图层
   */
  removeLayer(layer) {
    this._map.removeLayer(layer)
  }

  getBaselayers(layers) {
    if (_.isString(layers) || _.isObject(layers)) {
      return LayerPlugins.parse(layers, { tiandituToken: this.tiandituToken })
    }

    if (_.isArray(layers)) {
      const layerList = LayerPlugins.parse(layers, { tiandituToken: this.tiandituToken })

      layerList.forEach((i, index) => {
        i.setZIndex(index)
      })

      return layerList
    }

    if (_.isNil(layers)) return null

    throw new TypeError(`option.baselayers: ${layers} is not MapPlus.BaseLayer.MapList[?].Name 、 Array<MapPlus.BaseLayer.MapList[?].Name | Layer> or null`)
  }

  /**
   * 增加对应的基础的图层
   * @param {String} baseLayer 已经设置好的地图底图的名称
   */
  addBaseLayer(baseLayer, index) {
    let layer = null
    if (_.isString(baseLayer) || _.isObject(baseLayer)) {
      // TODO 应该添加哪里的图层是否存在的判断

      layer = LayerPlugins.parse(baseLayer, { tiandituToken: this.tiandituToken })
    } else {
      // 补充其他类型 ！！！
    }

    if (layer !== null) {
      this.baseLayers[index] = layer
      layer.setZIndex(index)
      this._map.getLayers().insertAt(index, layer)
    } else {
      // console.error('添加地图异常')
    }
  }

  /**
   * 移除对应的基础图层
   * @param {String} baseLayerName 已经设置好的地图底图的名称
   */
  removeBaseLayer(baseLayerName) {
    const layer = this._map.getLayers().getArray().find((_layer) => _layer.values_.id === baseLayerName)
    if (layer) this._map.removeLayer(layer)
  }

  /**
   * `addWMSLayer` 函数没想好怎么进行封装，先使用 `addGeoServerWMS3857Layer`
   */
  addGeoServerWMS3857Layer({
    serverURL,
    layers,
    visible = true,
    account,
    passwd,
    cqlFilter = null,
    styles = null,
    index = null,
  }, opt = {}) {
    if (_.isNumber(index)) opt.index = index

    const sourceOptions = {
      crossOrigin: 'anonymous',
      // 服务地址
      url: serverURL,
      // projection: new Projection(srs),
      params: {
        // 以下属性固定为大写，不允许进行修改
        'FORMAT': 'image/png', // eslint-disable-line quote-props
        'VERSION': '1.1.1', // eslint-disable-line quote-props
        'SRS': 'EPSG:3857', // eslint-disable-line quote-props
        layers,
        exceptions: 'application/vnd.ogc.se_inimage',
        // ratio,
      },
      // serverType,
      // projection,
    }

    if (cqlFilter) sourceOptions.params['CQL_FILTER'] = cqlFilter // eslint-disable-line dot-notation
    if (styles) sourceOptions.params['STYLES'] = styles // eslint-disable-line dot-notation

    if (_.isString(account)) sourceOptions.params.ACCOUNT = account
    if (_.isString(passwd)) sourceOptions.params.PASSWD = passwd

    const source = new ImageWMSSource(sourceOptions)

    const layer = new ImageLayer({
      visible,
      source,
    })

    this.addLayer(layer, opt)
    return layer
  }

  /**
   * 添加WMS服务
   */
  // addWMSlayer({
  //   serverURL,
  //   layers,
  //   projection = 'EPSG:3857',
  //   format = 'image/png',
  //   version = '1.1.1',
  //   visible = true,
  //   // ratio = 1,
  //   account,
  //   passwd,
  //   cqlFilter = null,
  //   styles = null,
  // }, opt = {}) {
  //   const params = {
  //     // 服务地址
  //     url: serverURL,
  //     // projection: new Projection(srs),
  //     params: {
  //       // 以下属性固定为大写，不允许进行修改
  //       'FORMAT': format, // eslint-disable-line quote-props
  //       'VERSION': version, // eslint-disable-line quote-props
  //       'SRS': projection, // eslint-disable-line quote-props
  //       layers,
  //       exceptions: 'application/vnd.ogc.se_inimage',
  //       // ratio,
  //     },
  //     // serverType,
  //     // projection,
  //   }

  //   if (cqlFilter) params.params['CQL_FILTER'] = cqlFilter // eslint-disable-line dot-notation
  //   if (styles) params.params['STYLES'] = styles // eslint-disable-line dot-notation

  //   if (!_.isNil(account)) params.params.ACCOUNT = account
  //   if (!_.isNil(passwd)) params.params.PASSWD = passwd

  //   const source = new ImageWMS(params)

  //   const layer = new ImageLayer({
  //     visible,
  //     source,
  //   })

  //   this.addLayer(layer, opt)

  //   return layer
  // }

  /**
   * 添加WMS Tile服务
   */
  // addWMSTilelayer({
  //   serverURL,
  //   layers,
  //   projection = 'EPSG:3857',
  //   format = 'image/png',
  //   version = '1.1.1',
  //   visible = true,
  //   // ratio = 1,
  //   account,
  //   passwd,
  //   cqlFilter = null,
  //   styles = null,
  // }, opt = {}) {
  //   const params = {
  //     // 服务地址
  //     url: serverURL,
  //     // projection: new Projection(srs),
  //     params: {
  //       // 以下属性固定为大写，不允许进行修改
  //       'FORMAT': format, // eslint-disable-line quote-props
  //       'VERSION': version, // eslint-disable-line quote-props
  //       'SRS': projection, // eslint-disable-line quote-props
  //       layers,
  //       exceptions: 'application/vnd.ogc.se_inimage',
  //     },
  //   }

  //   if (cqlFilter) params.params['CQL_FILTER'] = cqlFilter // eslint-disable-line dot-notation
  //   if (styles) params.params['STYLES'] = styles // eslint-disable-line dot-notation

  //   if (!_.isNil(account)) params.params.ACCOUNT = account
  //   if (!_.isNil(passwd)) params.params.PASSWD = passwd

  //   const source = new TileWMS(params)
  //   const layer = new TileLayer({
  //     visible,
  //     source,
  //   })

  //   this.addLayer(layer, opt)

  //   return layer
  // }

  addVectorLayer({
    index = null,
    style = null,
    visible = true,
  }) {
    const source = new VectorSource()
    const layer = new VectorLayer({
      zIndex: index,
      source,
      style,
      visible,
    })

    this.addLayer(layer, { index })

    return layer
  }

  /**
   * 添加WMTS服务
   */
  // addWMTSlayer({
  //   serverURL,
  //   layers,
  //   projection,
  //   format,
  //   visible,
  //   extent = [ -180.0, -90.0, 180.0, 90.0 ],
  // }, opt = {}) {
  //   const source = new WMTS({
  //     // 服务地址
  //     url: serverURL,
  //     layer: layers,
  //     // 切片集
  //     matrixSet: projection || 'EPSG:4326',
  //     format,
  //     projection: new Projection({
  //       code: 'EPSG:4326', // 投影编码
  //       units: 'degrees',
  //       axisOrientation: 'neu',
  //     }),
  //     // 切片信息
  //     tileGrid: new WMTSTileGrid({
  //       tileSize: [ 256, 256 ],
  //       extent, // 范围
  //       origin: [ -180.0, 90.0 ],
  //       resolutions,
  //       matrixIds: gridNames,
  //     }),
  //     wrapX: true,
  //   })

  //   // TODO 添加测试切片服务，增加加载相应时间，提高加载速度
  //   const layer = new TileLayer({
  //     source,
  //     visible,
  //   })

  //   // const source = new TileWMS(params)
  //   // const layer = new TileLayer({
  //   //   visible,
  //   //   source,
  //   // })

  //   this.addLayer(layer, opt)

  //   return layer
  // }

  /**
   * 添加WMTS服务
   */
  // addWMTS3857layer({
  //   serverURL,
  //   layers,
  //   projection,
  //   format = 'tiles',
  //   visible,
  //   extent = transformExtent([ -180.0, -90.0, 180.0, 90.0 ], 'EPSG:4326', 'EPSG:3857'),
  // }, opt = {}) {
  //   projection = getProjection('EPSG:3857') // eslint-disable-line
  //   const tileSizePixels = 256
  //   const tileSizeMtrs = getWidth(projection.getExtent()) / tileSizePixels
  //   const matrixIdsFor3857 = []
  //   const resolutionsFor3857 = []
  //   for (let i = 0; i <= 18; i += 1) {
  //     matrixIdsFor3857[i] = `EPSG:3857:${i}`
  //     resolutionsFor3857[i] = tileSizeMtrs / (2 ** i)
  //   }
  //   const tileGrid = new WMTSTileGrid({
  //     origin: getTopLeft(projection.getExtent()),
  //     resolutions: resolutionsFor3857,
  //     extent,
  //     matrixIds: matrixIdsFor3857,
  //   })

  //   const source = new WMTS({
  //     // 服务地址
  //     url: serverURL,
  //     layer: layers,
  //     // 切片集
  //     matrixSet: 'EPSG:3857',
  //     format,
  //     // matrixSet: 'w', // w 对应的为Web墨卡托投影
  //     // projection: new Projection({
  //     //   code: 'EPSG:3857', // 投影编码
  //     //   units: 'degrees',
  //     //   axisOrientation: 'neu',
  //     // }),
  //     // 切片信息
  //     tileGrid,
  //     // tileGrid: new WMTSTileGrid({
  //     //   tileSize: [ 256, 256 ],
  //     //   extent, // 范围
  //     //   // origin: transform([ -180.0, 90.0 ], 'EPSG:4326', 'EPSG:3857'),
  //     //   resolutions: resolutionsFor3857,
  //     //   matrixIds: gridNamesFor3857,
  //     // }),
  //     // wrapX: true,
  //   })

  //   // TODO 添加测试切片服务，增加加载相应时间，提高加载速度
  //   const layer = new TileLayer({
  //     source,
  //     visible,
  //   })

  //   // const source = new TileWMS(params)
  //   // const layer = new TileLayer({
  //   //   visible,
  //   //   source,
  //   // })

  //   this.addLayer(layer, opt)

  //   return layer
  // }

  /**
   * ------------------------------------------------
   * -------------------绘制模块----------------------
   * ------------------------------------------------
   */
  /**
   * 开始绘制
   * @param {String | null} style 绘制类型
   * @param { ol.style.Style } style 绘制的样式
   */
  startDrawTool(type = null, style = null) {
    if (_.isNull(this._drawTool)) this._drawTool = new DrawTool(this._map)
    this._drawTool.start(type, style)
  }

  startDrawToolOne(type = null, style = null, callback = null) {
    if (_.isNull(this._drawTool)) this._drawTool = new DrawTool(this._map)
    this._drawTool.startOne(type, style, callback)
  }

  destoryDrawTool() {
    this._drawTool.destory()

    this._drawTool = null
  }

  /**
   * 停止绘制
   * @param {Function} callback 回调函数
   */
  stopDrawTool(callback) {
    if (this._drawTool !== null) this._drawTool.stop(_.isFunction(callback) ? callback : () => { })
  }

  /**
   * 切换当前的绘制类型
   */
  changeDrawToolType(newType) {
    // TODO 在对外暴露地方增加格式验证功能
    this._drawTool.changeType(newType)
  }

  /**
   * 切换当前的绘制类型
   */
  drawTool4BackStep() {
    if (!_.isNull(this._drawTool)) this._drawTool.back()
  }

  /**
   * 清空当前绘制的features
   */
  cleanDrawToolFeatures() {
    if (!_.isNull(this._drawTool)) this._drawTool.clearFeaures()
  }

  /**
   * ------------------------------------------------
   * -------------------测量模块----------------------
   * ------------------------------------------------
   */

  /**
   * 开始测量
   */
  startMeasure(type = null, style = null) {
    if (_.isNull(this._measure)) this._measure = new Measure(this._map)
    this._measure.start(type, style)
  }

  /**
   * 停止测量功能
   */
  stopMeasure() {
    this._measure.stop()
  }

  /**
   * 切换当前的测量类型
   * @param {String} newType 新的测量类型
   */
  changeMeasureType(newType) {
    // TODO 在对外暴露地方增加格式验证功能
    this._measure.changeType(newType)
  }

  /**
   * ------------------------------------------------
   * -------------------事件绑定----------------------
   * ------------------------------------------------
   */
  bindClick(e) {
    this._map.on('singleclick', e)
  }

  // 双击事件
  bindDoubleClick(e) {
    this._map.on('dblclick', e)
  }

  bindMoveStart(e) {
    this._map.on('movestart', e)
  }

  bindMoveEnd(e) {
    this._map.on('moveend', e)
  }

  bindPointermove(e) {
    this._map.on('pointermove', e)
  }

  addArcGISLayer({
    url,
    projection = 'EPSG:4326',
    maxZoom = 18,
    wrapX = true,
    index = null,
  }) {
    const layer = new TileLayer({
      source: new XYZ({
        url,
        maxZoom,
        projection,
        wrapX,
      }),
    })

    this.addLayer(layer, { index })

    return layer
  }
}

export default MapPlus
